### PR TITLE
Add helpers for computing sums

### PR DESCRIPTION
**Pull request by richfitz**
Previous version of this was generated in odin.dust: https://github.com/mrc-ide/odin.dust/blob/master/R/generate_dust_sexp.R#L111-L138 or here for the 1d version https://github.com/mrc-ide/odin/blob/master/inst/library.c#L297-L303

### CODE REVIEW

File: inst/include/dust2/array.hpp [hunk 1]
```diff
@@ -85,5 +86,100 @@ size_t index(std::initializer_list<size_t> at,
...
+template <typename T, typename Container, typename BinaryOp>
+T reduce(Container x, const dimensions<1>& dim, T init, BinaryOp op,
+         const idx& i) {
+  if (i.first > i.second) {
+    return init;
+  }
+  return std::accumulate(x + i.first, x + i.second + 1, init, op);
+}
+
+// These are written out by hand for now, and then we an explore less
```
Comment (weshinsley) [inline inst/include/dust2/array.hpp:103]:
```suggestion
// These are written out by hand for now, and then we can explore less
```

File: inst/include/dust2/array.hpp [hunk 2]
```diff
@@ -85,5 +86,100 @@ size_t index(std::initializer_list<size_t> at,
...
+// tedious ways, and for more dimensions.  Previously we've generated
+// code for this rather than trying to do anything to clever with C++
```
Comment (weshinsley) [inline inst/include/dust2/array.hpp:105]:
```suggestion
// code for this rather than trying to do anything too clever with C++
```

**Review by weshinsley — APPROVED** [review]
OK - so just to check - 

We are insisting that for a:b in any of the sum dimensions, b>=a for our loops - which seems reasonable - and presumably there is no good reason why someone would want to sum the other way round...

**Comment by richfitz** [comment]
yeah - this is also the same for odin1, and we can look at relaxing (or erroring) later if this is not met

**Pull request merged**